lib/Dialect/HW/HWDialect.cpp

```diff
@@ -57,6 +58,27 @@ struct HWOpAsmDialectInterface : public OpAsmDialectInterface {
 };
 } // end anonymous namespace
 
+namespace {
+/// This class defines the interface for handling inlining with HW operations.
+struct HWInlinerInterface : public mlir::DialectInlinerInterface {
+  using mlir::DialectInlinerInterface::DialectInlinerInterface;
+
+  bool isLegalToInline(Operation *op, Region *, bool,
+                       BlockAndValueMapping &) const final {
+    return isa<ConstantOp>(op) || isa<BitcastOp>(op) ||
+           isa<ArrayCreateOp>(op) || isa<ArrayConcatOp>(op) ||
+           isa<ArraySliceOp>(op) || isa<ArrayGetOp>(op) ||
+           isa<StructCreateOp>(op) || isa<StructInjectOp>(op) ||
+           isa<UnionCreateOp>(op) || isa<UnionExtractOp>(op);
+  }
+
+  bool isLegalToInline(Region *, Region *, bool,
+                       BlockAndValueMapping &) const final {
+    return false;
+  }
+};
+} // end anonymous namespace
+
```

Comment:
I think this looks good to me, but I'm not an expert on how the inlining works and if there are any implications of this code. But since we haven't used the inliner with HW yet, this looks like a first solid step -- and the tests are passing.

Reply:
Yes, I try to be defensive here, to get the tests working I need `ConstantOp` to be inlineable. The other operations are also used by LLHD (except union) and may occur in functions thus I also allow them, but nothing else. I.e., those operations may be inlined when occuring in a standard function, but no `hw.module`, `llhd.proc` or such would be inlined.